app/services/corrections/CorrectionService.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services.corrections
+
+import cats.implicits._
+import models.{DataMissingError, Period, UserAnswers, ValidationResult}
+import models.corrections.PeriodWithCorrections
+import models.requests.corrections.CorrectionRequest
+import pages.corrections.CorrectPreviousReturnPage
+import queries.corrections.AllCorrectionPeriodsQuery
+import services.PeriodService
+import uk.gov.hmrc.domain.Vrn
+
+import java.time.LocalDate
+import javax.inject.Inject
+
+class CorrectionService @Inject()(
+                                   periodService: PeriodService
+                                 ) {
+
+  def fromUserAnswers(answers: UserAnswers, vrn: Vrn, period: Period, commencementDate: LocalDate): ValidationResult[CorrectionRequest] = {
+
+    if (firstPeriod(period, commencementDate)) {
+      CorrectionRequest(vrn, period, List.empty).validNec
+    } else {
+      getCorrections(answers).map { corrections =>
+        CorrectionRequest(vrn, period, corrections)
+      }
+    }
+  }
+
+  private def firstPeriod(period: Period, commencementDate: LocalDate): Boolean = {
+    periodService.getReturnPeriods(commencementDate).headOption match {
+      case Some(firstAvailableReturnPeriod) => firstAvailableReturnPeriod == period
+      case _ => false
+    }
+  }
+
+  private def getCorrections(answers: UserAnswers): ValidationResult[List[PeriodWithCorrections]] =
+    answers.get(CorrectPreviousReturnPage) match {
+      case Some(false) =>
+        List.empty[PeriodWithCorrections].validNec
+      case Some(true) =>
+        processCorrections(answers)
+      case None =>
+        DataMissingError(CorrectPreviousReturnPage).invalidNec
+    }
+
+  private def processCorrections(answers: UserAnswers): ValidationResult[List[PeriodWithCorrections]] = {
+    answers.get(AllCorrectionPeriodsQuery) match {
```

Comment:
I think we might need to validate at a lower level as well to check that all data inside each correction exists aswell, this will be where we will catch things like a user having said they want a correction, then choosing the country, then going back and not entering it. I think that would pass this validation, but there would be a correction amount missing? Let me know if I have mis-interpreted this.